prover/src/matrix/row_matrix.rs

```diff
@@ -0,0 +1,155 @@
+    pub fn new(data: Vec<E>, row_len: usize) -> Self {
+        assert!(data.len() % row_len == 0);
+        assert!(row_len % ARR_SIZE == 0);
+        assert!(!data.is_empty());
+
+        Self { data, row_len }
+    }
+
+    /// Converts a column-major matrix of polynomials `Matrix<E>` into a RowMatrix evaluated at
+    /// a shifted domain offset.
+    pub fn from_polys(polys: &Matrix<E>, blowup_factor: usize) -> Self {
+        // get the number of rows and columns in the polys.
+        let row_width = polys.num_cols();
+        let num_rows = polys.num_rows();
+
+        // get the twiddles for the segment.
+        let twiddles = fft::get_twiddles::<E::BaseField>(polys.num_rows() * blowup_factor);
+        let domain_offset = E::BaseField::GENERATOR;
+
+        // get the number of segments in the matrix.
+        let num_of_segments = row_width / ARR_SIZE;
+        let mut segments = Segments::new(Vec::new());
+
+        // precompute offsets for each row.
+        let mut offsets = Vec::with_capacity(num_rows);
+        offsets.push(E::BaseField::ONE);
+        for i in 1..num_rows {
+            offsets.push(offsets[i - 1] * domain_offset);
+        }
+
+        // create segments.
+        (0..num_of_segments).for_each(|i| {
+            // create a vector of arrays to hold the result.
+            let mut result_vec_of_arrays =
+                unsafe { uninit_vector::<[E; ARR_SIZE]>(num_rows * blowup_factor) };
+
+            // transpose the segment into `Segment` and evaluate the polynomials at the
+            // shifted domain offset.
+            (i * ARR_SIZE..(i + 1) * ARR_SIZE).for_each(|col_idx| {
+                let row = polys.get_column(col_idx);
+                row.iter().enumerate().for_each(|(row_idx, elem)| {
+                    result_vec_of_arrays[row_idx][col_idx - i * ARR_SIZE] =
+                        elem.mul_base(offsets[row_idx]);
+                });
+            });
+
+            // create a `Segment` object from the result.
+            let row_matrix = Segment::new(result_vec_of_arrays);
+            segments.push(row_matrix);
+        });
+
+        // evaluate the polynomials in each segment at the shifted domain offset.
+        for segment in segments.iter_mut() {
+            segment.evaluate_poly(&twiddles);
+        }
+
+        let num_rows = num_rows * blowup_factor;
+
+        // create a vector of arrays to hold the result.
+        let mut result = unsafe { uninit_vector::<[E; ARR_SIZE]>(num_rows * row_width / ARR_SIZE) };
+
+        // transpose the segments into a row matrix.
+        segments.iter().enumerate().for_each(|(i, segment)| {
+            (segment.as_data()).iter().enumerate().for_each(|(j, row)| {
+                result[j * row_width / ARR_SIZE + i] = *row;
+            })
+        });
+
+        // create a `RowMatrix` object from the result.
+        RowMatrix {
+            data: flatten_vector_elements(result),
+            row_len: row_width,
+        }
```

Comment:
I would encapsulate this into a stand-alone function.

Also, a comment for the future: I wonder if we could improve it by using some kind of in-place transposition algorithm to avoid additional memory allocation.

Reply:
After considering the problem, I believe that transposing the data in place is not a straightforward task. Consider the desired final layout of the data to be in the format(location of array in the final vector) `vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]`, but we receive `vec![1, 5, 9, 2, 6, 10, 3, 7, 11, 4, 8, 12]` from the current Segment's approach.

I don't at the moment have a solution that allows us to directly achieve the final transposed state in just one pass without allocating new memory. I'll think about it but for the time being I'm sending an update using the current approach.

Reply:
I've added a [TODO](https://github.com/facebook/winterfell/blob/934f0db10a959f08490f9d8cb09281a5fbe64425/prover/src/matrix/row_matrix.rs#L90).